resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/PreferNonAllocApiAnalyzer.cs

```diff
@@ -0,0 +1,109 @@
+using System.Linq;
+using JetBrains.Annotations;
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Errors;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Dispatcher;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.ExtensionsAPI.Resolve.Filters;
+using JetBrains.ReSharper.Psi.Resolve;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Analysis
+{
+    [ElementProblemAnalyzer(typeof(IInvocationExpression), HighlightingTypes = new[] {typeof(PreferNonAllocApiWarning)})]
+    public class PreferNonAllocApiAnalyzer : UnityElementProblemAnalyzer<IInvocationExpression>
+    {
+        public PreferNonAllocApiAnalyzer([NotNull] UnityApi unityApi) : base(unityApi)
+        {
+        }
+
+        protected override void Analyze(IInvocationExpression expression, ElementProblemAnalyzerData data,
+            IHighlightingConsumer consumer)
+        {
+            if (!(expression.InvokedExpression is IReferenceExpression referenceExpression)) 
+                return;
+
+            var reference = expression.Reference;
+            if (reference == null) 
+                return;
+            
+            var info = reference.Resolve();
+            if (info.ResolveErrorType == ResolveErrorType.OK && info.DeclaredElement is IMethod allocMethod)
+            {
+                var nonAllocMethod = GetNonAllocVersion(allocMethod, expression);
+                if (nonAllocMethod != null)
+                    consumer.AddHighlighting(new PreferNonAllocApiWarning(expression, referenceExpression, nonAllocMethod));
+            }
+        }
+
+        private static IMethod GetNonAllocVersion([NotNull]IMethod method,[NotNull] IInvocationExpression expression)
+        {
+            var originName = method.ShortName;
+            
+            string newName; // xxx[All] -> xxxNonAlloc
+            if (originName.EndsWith("All"))
+            {
+                newName = originName.Substring(0, originName.Length - 3) + "NonAlloc";
+            }
+            else
+            {
+                newName = originName + "NonAlloc";
+            }
```

Comment:
Move this section below the check for type, or we'll be doing this for every method invocation in the project.